### PR TITLE
Support pattern matching on constants

### PR DESCRIPTION
**Pull request by hubertp**
### Pull Request Description

This change adds support for matching on constants by:
1) extending parser to allow literals in patterns
2) generate branch node for literals

Related to https://www.pivotaltracker.com/story/show/182743559

### Checklist

Please include the following checklist in your PR:

- [ ] The documentation has been updated if necessary.
- [x] All code conforms to the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [ ] If GUI codebase was changed: Enso GUI was tested when built using BOTH
        `./run ide build` and `./run ide watch`.

### CODE REVIEW

File: engine/runtime/src/main/java/org/enso/interpreter/node/controlflow/caseexpr/LiteralBranchNode.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    return LiteralBranchNodeGen.create(literal, branch);
+  }
+
+  @Specialization
+  void doText(
+      VirtualFrame frame,
+      Object state,
+      Text target,
+      @Cached("build()") ToJavaStringNode toJavaStringNode) {
+    if (textProfile.profile(literal.equals(toJavaStringNode.execute(target))))
```
Comment (radeusgd) [inline engine/runtime/src/main/java/org/enso/interpreter/node/controlflow/caseexpr/LiteralBranchNode.java]: This sounds like a possible overkill, but I'm wondering if we should use the same equality as our `==` operator, i.e. one that handles Unicode normalization. OTOH maybe it makes sense for pattern matching to be 'more exact'.

**Review by radeusgd — APPROVED** [review]
Awesome! I can't wait to use it :smiley:

File: engine/runtime/src/main/java/org/enso/interpreter/node/controlflow/caseexpr/LiteralBranchNode.java (hunk 1 above)
Comment (JaroslavTulach) [inline engine/runtime/src/main/java/org/enso/interpreter/node/controlflow/caseexpr/LiteralBranchNode.java]: Converting `Text` to `String` just to do `equals`? I guess we could benefit from a switch to [TruffleStrings](https://github.com/oracle/graal/blob/master/truffle/docs/TruffleStrings.md) then and use [EqualsNode](https://www.graalvm.org/truffle/javadoc/com/oracle/truffle/api/strings/TruffleString.EqualNode.html) which would compare the values without converting to `String`.

File: engine/runtime/src/main/java/org/enso/interpreter/node/controlflow/caseexpr/LiteralBranchNode.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+@NodeInfo(shortName = "LiteralMatch", description = "Allows matching on literals")
+public abstract class LiteralBranchNode extends BranchNode {
+  private final Object literal;
+
+  private final ConditionProfile textProfile = ConditionProfile.createCountingProfile();
+  private final ConditionProfile numProfile = ConditionProfile.createCountingProfile();
+
+  LiteralBranchNode(Object literal, RootCallTarget branch) {
+    super(branch);
```
Comment (JaroslavTulach) [inline engine/runtime/src/main/java/org/enso/interpreter/node/controlflow/caseexpr/LiteralBranchNode.java]: What are the actual values of `literal`? `String`, `long`, `EnsoBigInteger`? Should they be enforced? By `assert`? By multiple variants of the constructor?

File: engine/runtime/src/test/scala/org/enso/compiler/test/codegen/AstToIrTest.scala [hunk 3]
```diff
@@ -213,6 +213,38 @@ class AstToIrTest extends CompilerTest with Inside {
...
+
+      ir.branches(0).pattern shouldBe an[IR.Pattern.Literal]
+      ir.branches(1).pattern shouldBe an[IR.Pattern.Literal]
+    }
+
+    "support constructor patterns with nested literals" in {
+      val ir =
+        """
+          |case foo of
+          |    Cons 1 b -> a + b
```
Comment (JaroslavTulach) [inline engine/runtime/src/test/scala/org/enso/compiler/test/codegen/AstToIrTest.scala:233]: Interesting that `Cons 1 b` works!

File: test/Tests/src/Semantic/Case_Spec.enso [hunk 4]
```diff
@@ -88,5 +88,18 @@ spec = Test.group "Pattern Matches" <|
         case Any of
             Any -> Nothing
             _ -> Test.fail "Expected the Any constructor to match."
+    Test.specify "should be able to match on literal values" <|
```
Comment (JaroslavTulach) [inline test/Tests/src/Semantic/Case_Spec.enso:91]: Shouldn't the `.enso` test also check for `Cons 1 b` style?

**Review by JaroslavTulach — APPROVED** [review]
Great work in a really fast time!

File: engine/runtime/src/test/scala/org/enso/compiler/test/codegen/AstToIrTest.scala (hunk 3 above)
Comment (hubertp) [reply engine/runtime/src/test/scala/org/enso/compiler/test/codegen/AstToIrTest.scala:233]: It parses, it doesn't necessarily work :p

File: engine/runtime/src/main/java/org/enso/interpreter/node/controlflow/caseexpr/LiteralBranchNode.java (hunk 1 above)
Comment (hubertp) [reply engine/runtime/src/main/java/org/enso/interpreter/node/controlflow/caseexpr/LiteralBranchNode.java]: Yeah, this is the only ugly part in this PR, I admit. 
Thanks for the suggestion, I wasn't entirely sure what is _the proper way_ to deal with this.

**Review by 4e6 — APPROVED** [review]

File: engine/runtime/src/main/java/org/enso/interpreter/node/controlflow/caseexpr/LiteralBranchNode.java (hunk 1 above)
Comment (hubertp) [reply engine/runtime/src/main/java/org/enso/interpreter/node/controlflow/caseexpr/LiteralBranchNode.java]: On the other hand, we use similar technique already with `ExpectStringNode`, `ExpectTextNode` (and `ToJavaStringNode`) in various places.

File: engine/runtime/src/main/java/org/enso/interpreter/node/controlflow/caseexpr/LiteralBranchNode.java (hunk 2 above)
Comment (hubertp) [reply engine/runtime/src/main/java/org/enso/interpreter/node/controlflow/caseexpr/LiteralBranchNode.java]: done

File: engine/runtime/src/test/scala/org/enso/compiler/test/codegen/AstToIrTest.scala (hunk 3 above)
Comment (hubertp) [reply engine/runtime/src/test/scala/org/enso/compiler/test/codegen/AstToIrTest.scala:233]: And now it works

File: test/Tests/src/Semantic/Case_Spec.enso (hunk 4 above)
Comment (hubertp) [reply test/Tests/src/Semantic/Case_Spec.enso:91]: Yeap, now it does

File: test/Tests/src/Semantic/Case_Spec.enso [hunk 5]
```diff
@@ -101,5 +101,13 @@ spec = Test.group "Pattern Matches" <|
...
             "foo" -> Nothing
             _     -> Test.fail "Expected value to match constant."
+    Test.specify "should be able to match on literal values nested in constructors" <|
+        value_1 = Cons 42 Nil
+        case value_1 of
+            42                    -> Test.fail "Expected value to match constant."
+            Cons (Cons 1 Nil) Nil -> Test.fail "Expected value to match constant."
+            Cons 1 Nil            -> Test.fail "Expected value to match constant."
+            Cons 42 Nil           -> Nothing
+            _                     -> Test.fail "Expected value to match constant."
```
Comment (radeusgd) [inline test/Tests/src/Semantic/Case_Spec.enso]: Maybe would be good to also add a test where one field is a literal but another is a binder? i.e. `Cons 42 foo ->` or `Cons 42 _ ->` ? Just for the sake of completeness.

File: engine/runtime/src/main/java/org/enso/interpreter/node/controlflow/caseexpr/LiteralBranchNode.java (hunk 1 above)
Comment (hubertp) [reply engine/runtime/src/main/java/org/enso/interpreter/node/controlflow/caseexpr/LiteralBranchNode.java]: Done

Comment (hubertp) [reply engine/runtime/src/main/java/org/enso/interpreter/node/controlflow/caseexpr/LiteralBranchNode.java]: So I think that with a comparison of normalized strings this is no longer an option.

File: test/Tests/src/Semantic/Case_Spec.enso (hunk 5 above)
Comment (hubertp) [reply test/Tests/src/Semantic/Case_Spec.enso]: added

**Pull request merged**